### DISCUSSION TITLE
Improving Support for Dependency Versions

### DISCUSSION THREAD

**Discussion by pyrmont** [janet-lang/janet#1264] [Ideas]
One of the limitations of Janet's code importing system is dependency versioning. Although a user can specify in a `project.janet` file a Git tag/commit to use, there is no guarantee that the Janet interpreter will import the file from this revision of the repository. This discussion proposes a backwards-compatible way in which Janet could allow users to specify a dependency at a particular revision.

## Background

Janet source files can import other Janet source files (each, a 'dependency'). When a dependency is imported, the source file is read, parsed and compiled into values that are stored in an environment table. During importation, the public values of this table are merged into the environment table of the importing file (with the symbols typically prepended with a name so as to avoid conflicts). A Janet source file can import dependencies that are located relative to the importing file or it can import dependencies from the system library (often `/usr/local/lib/janet`).

Janet files are typically added to the system library using JPM, the Janet Project Manager. By listing the URLs of Git repositories in a `project.janet` file, a user can import the dependencies into the system library using `jpm deps`. JPM allows a user to import the files at a particular revision of each repository. Users do this by specifying a `:tag` (despite the name, the value can be a Git tag or a commit SHA).

## Problem

Unfortunately, while a user can specify a revision in the `project.janet` file, there's no guarantee that the dependency imported at runtime is the version from the revision specified. If the user subsequently installs a different version of the file, this will simply overwrite the version in the system library.

## Solution

This post proposes a two-part solution. The proposal is detailed below but the way this would be called by the user is demonstrated first.

Rather than this:

```janet
(import foo)

(foo/bar)
```

a user could write this:

```janet
(import foo :tag "1.2.3")

(foo/bar)
```

In other words, the user would specify at the call site of `import` the revision desired. Janet would then import a version of the file from that revision using the tag as part of the filename (e.g. `foo.1.2.3.janet`). If the file with that name could not be located, an error would be thrown. Existing code would work exactly the same as now; specifying a tag would be optional.

## Implementation

As noted above, the solution involves two parts.

### Janet interpreter

The first part is a change to the private `require-1` function in Janet's `boot.janet`.

```diff
(defn- require-1
   [path args kargs]
+   (when (and (string/has-prefix? "." path) (dyn :tag))
+     (put kargs :tag (dyn :tag)))
+   (def apath (if (kargs :tag) (string path "." (kargs :tag)) path))
+   (def [fullpath mod-kind] (module/find apath))
-   (def [fullpath mod-kind] (module/find path))
   (unless fullpath (error mod-kind))
   (if-let [check (if-not (kargs :fresh) (in module/cache fullpath))]
     check
    (if (module/loading fullpath)
      (error (string "circular dependency " fullpath " detected"))
       (do
         (def loader (if (keyword? mod-kind) (module/loaders mod-kind) mod-kind))
         (unless loader (error (string "module type " mod-kind " unknown")))
+         (def newenv (merge-into (make-env) {:tag (kargs :tag)}))
+         (def env (loader fullpath (array/concat @[:env newenv] args)))
-         (def env (loader fullpath args))
         (put module/cache fullpath env)
         env))))
```

The changes here:

1. Set the `(kargs :tag)` to the value of a dynamic binding if the import is for a relative path (this dynamic binding is set up later).

2. Append `.` and the value of `(kargs :tag)` to the path if a value exists.

3. Search for the modified path.

4. Set the initial environment for the imported file to include the `(kargs :tag`) value as a dynamic binding (this is necessary for the relative paths described at 1).

No other changes are required to the interpreter.

### JPM

The second part involves changes to JPM. These would cause JPM to append the tag specified in a `project.janet` to the Janet source files installed into the system library.

A symlink (without the tag) could also be created to point to the source file. This would allow the source file to be used by import statements that don't include a tag.

### Inconsistent with other languages

Some may object that specifying the tag at the import call site is inconsistent with other programming languages. While I agree this is true, I argue that this is the preferable approach.

The author of the source file is in the best position to know which revision of the dependency they wish to use and they should be able to specify this explicitly. It should not be the case that a user of the source file can change the revision of the dependencies used (without any change to the source code). We should avoid unnecessary coupling but the revision of a dependency is _necessarily_ coupled to the file importing it. A different revision of a dependency may change the implementation of a function used or it may change the function signature, both of which may be incompatible. By allowing an author to specify the revision, that author can know the code will not run unless that specific revision is available.

### Painful to update dependency versions

When updating a dependency, a user would need to change both the `project.janet` file as well as the import call sites. I agree this is painful but argue that it is a necessary consequence of being explicit about the revision used.

### Increased resources

This increases the disk space used, particularly as more and more Janet source files are installed in the system library. My argument would be that, in 2023, this is fine.

### Non-source dependencies

This proposal has only discussed Janet source files but Janet also supports native dependencies and Janet image files. The way the code has been suggested above would allow a user to specify a tag for a native dependency.

While I don't think that would cause an issue if only one version of that dependency is imported, I don't think multiple revisions of native dependencies can be imported into the same process (but maybe they can—I'm not sure). I'm also unsure what the consequences of multiple Janet image files would be.

### `init.janet` files

The above proposal doesn't work cleanly for source files that are found by appending `/init.janet` to the path. I don't have a great answer for this at the moment.

## Thoughts

For those who read through all that, I'm curious what you think. Any thoughts? :)

**Comment by ianthehenry** [comment]
I think that the problem you're describing is already solved pretty well by project-local dependency directories? If I'm understanding correctly, you only have this clobbering issue if you install dependencies to the system path, and if you don't use the system path the issue goes away for free.

> The author of the source file is in the best position to know which revision of the dependency they wish to use and they should be able to specify this explicitly. It should not be the case that a user of the source file can change the revision of the dependencies used (without any change to the source code). We should avoid unnecessary coupling but the revision of a dependency is necessarily coupled to the file importing it. A different revision of a dependency may change the implementation of a function used or it may change the function signature, both of which may be incompatible. By allowing an author to specify the revision, that author can know the code will not run unless that specific revision is available.

If I'm distributing an application, I want to specify exactly the version of all of my dependencies that I expect. But if I'm distributing a library, I *don't* want to set exact constraints on what will become transitive dependencies for applications or libraries that use my library.

If library `foo` depends on some particular version of `spork`, and library `bar` depends on a *different* version of `spork`, then I can't write an application that depends on both `foo` and `bar`, and I'm sad (assuming that the version differences are trivial and the two versions are not actually incompatible).

If I'm the author of `foo` and `bar`, I really want to set constraints on the range of versions of `spork` that I need. (Which assumes that `spork` were using something like semver, which, you know, it's not even versioned, so that's maybe not the best example.)

I know that JPM gives me no way to express this right now, but that makes me sad, and I wouldn't want to get further away from what I see as the right end goal.

> Painful to update dependency versions

I would add another part to this, which is that it's a bit fragile: you could update the version in your `project.janet` file without updating every `use` and `import` and `require` in your codebase, and your code would continue to work fine -- on your machine. But if installed it from scratch, it wouldn't work.

> Thoughts

It seems like a heavy solution to a problem that already has a cheap solution. Maybe it's worth saying more about what you're using the system path for?

I think that Janet's packaging story has a lot of shortcomings (dev dependencies! -- I know you're aware of that one :) and there's certainly a lot of room for improvement in JPM. I haven't felt the pain that this particular proposal allays, and I think that you could get the same benefit with no change to user code with a more ambitious change to JPM's library directory structure. Which *would* be a backwards-incompatible change to existing installations, but would be completely transparent to the user...

**Reply by pyrmont** [reply]
> I think that the problem you're describing is already solved pretty well by project-local dependency directories?

This is true for a user who is developing an application since they can simply install everything local to the project but falls apart as soon as you try to distribute the application. You either have to compile the program so that the dependencies are all baked in (which only solves the problem if you ship the actual binaries to users and now you need to cross-compile for all major architectures) or you have to ship an entire directory and tell your users to run that (somehow). Neither option works well with JPM.

> But if I'm distributing a library, I _don't_ want to set exact constraints on what will become transitive dependencies for applications or libraries that use my library.

People say this all the time but I confess it makes no sense to me. Whether I'm developing an application or a library, if I have dependencies, the code is coupled. Unless you're talking about no-dependency libraries (in which case, this proposal is irrelevant), the library author is depending on code with particular characteristics. I think those should be explicit.

> If library `foo` depends on some particular version of `spork`, and library `bar` depends on a _different_ version of `spork`, then I can't write an application that depends on both `foo` and `bar`, and I'm sad (assuming that the version differences are trivial and the two versions are not actually incompatible).

I probably should have mentioned this explicitly but my understand of how Janet works is that this proposal _would_ allow multiple versions of the same library. Since importing merely involves the creation of an environment table that's cached under a full file system path, and since the full file system paths for different versions are different under this proposal, you'd have no issue depending on different versions of Spork. They'd all coexist happily in different environment tables.

> It seems like a heavy solution to a problem that already has a cheap solution. Maybe it's worth saying more about what you're using the system path for?

I guess it seems to me that the system library is a giant foot gun and it bothers me that it exists in such a state. It either should be eliminated or it should be fixed. What is it for if we tell everyone never to use it?

I realise this is true for basically every dynamic language and the way that almost everyone (except Node) solves it is to create a second tool (e.g. Ruby's Bundler) that does all this magic to make it so that you only load particular versions of a dependency. I am occupying a pretty lonely position but my argument is that this is a bad design. It's a design fit for an interpreter that doesn't know users primarily import libraries written by other people. It couples the interpreter to this secondary tool which is bad because this coupling shouldn't be necessary but is because of the limitations of the interpreter. It also mean you can't have multiple versions of a library in use (even when, as in Janet's case, the interpreter itself can handle this just fine).

**Reply by ianthehenry** [reply]
> This is true for a user who is developing an application since they can simply install everything local to the project but falls apart as soon as you try to distribute the application. You either have to compile the program so that the dependencies are all baked in (which only solves the problem if you ship the actual binaries to users and now you need to cross-compile for all major architectures) or you have to ship an entire directory and tell your users to run that (somehow). Neither option works well with JPM.

That's true! But I think that the "compile for every architecture" is a very good option, because it allows you to distribute your programs to people who don't even have Janet or JPM installed in the first place -- a pretty large segment of the population. It also lets you freeze not only your Janet dependencies, but also the version of Janet that you depend on -- which does not make a very strong guarantee of backwards compatibility.

I guess I see distributing applications as a problem solved by existing system package managers, and I'm sad when I see a tool that asks me to install it with `go get` or `pip` or something -- I don't want to install language tooling for a language I don't know just to then install the thing I actually care about.

> People say this all the time but I confess it makes no sense to me. Whether I'm developing an application or a library, if I have dependencies, the code is coupled. Unless you're talking about no-dependency libraries (in which case, this proposal is irrelevant), the library author is depending on code with particular characteristics. I think those should be explicit.

> Since importing merely involves the creation of an environment table that's cached under a full file system path, and since the full file system paths for different versions are different under this proposal, you'd have no issue depending on different versions of Spork. They'd all coexist happily in different environment tables.

This is very true if you're depending on a library as a completely private, internal dependency -- it's fine to install multiple versions of that dependency side-by-side, and apart from some theoretical inefficiency it doesn't matter if you load multiple versions of those libraries at once.

But if I write a library that exposes a function that returns a Spork `tarray`, and another library exposes a function that takes a Spork `tarray`, it's important that an application depending on both libraries only builds one version of Spork.

I've never seen a package manager try to distinguish these two cases! But it seems like it would be a perfectly reasonable thing to do. "Internal" and "external" dependencies, or "private" and "public" dependencies or something like that. All the package managers I have used make the conservative assumption that all of a library's dependencies are "public," and thus must be consistent with one another -- even though I expect that the majority of library dependencies in practice are of the "private" variety!

> I guess it seems to me that the system library is a giant foot gun and it bothers me that it exists in such a state. It either should be eliminated or it should be fixed. What is it for if we tell everyone never to use it?

I think of the global package cache as "the packages that are available when I don't have a `project.janet` file." That is, when I'm opening the REPL to do some exploratory programming, or when I'm writing a little shebang script. I think that it's a very big footgun that JPM operates on the system path by default -- it seems to be encouraging people to put more things in there -- but I think that "default packages for my personal scripts" is a useful enough thing to have that I would be sad to eliminate entirely.

**Reply by sogaiu** [reply]
If the following is the intent (or one of the intents) of the cache:

> I think of the global package cache as "the packages that are available when I don't have a project.janet file." That is, when I'm opening the REPL to do some exploratory programming, or when I'm writing a little shebang script.

I think it would make sense for this to be stated in the docs explicitly -- AFAIU I don't think it is.

Seems far too easy to end up with broken things though.  I have had breakage multiple times and it hasn't been the case that detection of issues always occurs close to the time of breakage.

**Reply by CosmicToast** [reply]
> I guess I see distributing applications as a problem solved by existing system package managers

Just figured I would chime in, given that I've a long history in distro development.
System package managers exist in the state that they do because C's ecosystem is a trainwreck.
There is no other reason for their existence, it's not for convenience or anything like that.
It's because managing C (and especially C DSO versions and the dependency hell that ensues) is nearly impossible otherwise.

This isn't to say that, for instance, `pip` or `cargo` are the epitome of good design - far from it, but they have their own problems.
Application distribution is fundamentally an unsolved question, which I obviously have strong opinions on, but I assure you that, for instance, go libraries, do not really belong in system package managers.
Attempts to manage python dependencies are more commonplace (mostly due to tools like dnf4, portage, etc being written in python), and are often incomplete and/or incompatible with developmental workflows (paradoxically, despite portage being written in python, python is one of the more fragile languages to work with on gentoo, for instance).

Anyway, as for multiple library version dependencies across a system, that *is* a solved problem (though there are around 3 different solutions, all with their own tradeoffs).
The proposed one is (as far as I understand) is to allow parallel version installs in the global package cache. This is an approach commonly taken by C (and thus package managers, at least those that allow multiple simultaneous versions, such as portage and nix).
I personally think it's an acceptable solution, involving minimal intervention.

For completeness purposes, the other solutions involve:
* A project-local directory, ala node. Note, however, that the node ecosystem is moving away from this gradually, via things like pnpm and deno, which utilize the third option. This would allow distributing the entire development directory together per-architecture if you really needed a standalone distributable.
* A user-local id cache which includes the version/sha in the descriptor, and can be resolved. This is becoming progressively more common after being pioneered by go. This does have a distribution disadvantage, however, unless you can produce static (as far as the language itself is concerned, not counting external DSOs) distributables. Otherwise you require everything to be present at runtime (as with the new node offerings, though deno handles this by downloading them on first runtime.

The issue is that currently, jpm is rather broken when attempting to produce static binaries (I've been vocal about this previously), but I've lost quite a lot of motivation to work on things like that lately (on top of my already restricted free time due to irl circumstances).

Either way, just figured I'd chime in with some context, because from a distrodev perspective the conversation was getting a bit weird.
Hope it was useful!

**Reply by bakpakin** [reply]
JPM does need a bit of a rewrite (and should be simpler), but I would disagree with some of the issues raised:

* Do not look to node/deno for package management influence. The only reason they _seem_ to work well is that nobody uses native packages that interface with the OS, which means the whole problem is trivial. All of the problems that these package managers "solve" are self-inflicted.
* I don't really want to support multiple versions of the same package in a tree. What _might_ make sense is the ability to install packages to a different location like installing spork to `subtree/spork` and being able to import that separately, no need for tags or redoing the import mechanism.
* Dev dependencies aren't needed in languages that can compile and "tree-shake". The whole concept is because in the node world, people use to actually run the entire source tree as the distributed application. The idea with jpm is that you can build an image or a static library and distribute that instead. I know people will disagree with me on this, but I really think people are just stuck in the python/node mindset of "copy directory to destination to deploy/distribute".
* Default linking patterns for static builds should be fixed, although I think that these issues can be fixed (are fixed?) in JPM as is. I don't run into problems like this although other users seem to.

Probably a number of other issues I could think of but these are the biggest in my mind.

**Reply by bakpakin** [reply]
As far as fixing linking/building, I have been working on a port of `jpm/cc.janet` to spork [here](https://github.com/janet-lang/spork/blob/master/spork/cc.janet). This should hopefully help address both linking issues with system packages, allow users to compile stuff without needing all of the jpm mechanisms, be easier to configure, yet still have enough of an opinionated take to actually provide some benefit. We can also use some of the other functionality in spork to make things easier/cleaner.

**Reply by CosmicToast** [reply]
Looks like a very good start.
What's the recommended way to use this?
Let's say I want to produce an executable from `script.janet` which imports `jurl` (which contains `jurl/native`, a native module which depends on a pkgconfig package and `jurl/*` which are janet modules).
As far as I'm aware, you cannot expect a given package to be present when executing jpm as of currently, since jpm is what would normally be used to install it (so a clean janet+jpm install + clone + jpm command wouldn't have, for instance, spork available).

**Reply by bakpakin** [reply]
Yeah, this isn't a full build system, just a wrapper around the compiler. Notice also the msvc is a separate interface because it works differently enough. I don't recommend using this right now with jpm - this is something that could/will be used to _replace_ jpm. While you could plumb it together, I don't really see the point.

> As far as I'm aware, you cannot expect a given package to be present when executing jpm as of currently, since jpm is what would normally be used to install it (so a clean janet+jpm install + clone + jpm command wouldn't have, for instance, spork available).

This is currently done in jpm with the `post-deps` macro in project.janet, which is just a block of code that isn't executed for certain subcommands (like `jpm deps`), allowing some functionality to be used even before dependencies are installed.

**Reply by bakpakin** [reply]
As far as how to use this, to create a Janet native module (on posix platforms), the following should work (using the file json.c in spork):

```
(import spork/cc)
(setdyn cc/*visit* cc/visit-execute-if-stale)
(cc/search-static-libraries "m" "rt" "dl")
(cc/compile-and-link-shared "json.so" "json.c")
```

A build system could do more, for example cache whether or not all of the runtime libraries are needed, build both a static and dynamic version, but this is sufficient and will work in a REPL. Also, note that this _should_ work on BSDs that don't have `-lm` or `-lrt` and will not try to link to them (I haven't actually tested this), but the idea is to auto-detect available libraries, and return any that are missing. The user can then look for dynamic versions of these libraries, or raise and error, or try to install from a system package manager, or do whatever.

**Reply by sogaiu** [reply]
The last step didn't work for me.  Looks like `janet.h` is not being picked up.

Below is a log of steps taken with output.

```
$ jpm install spork
remote: Enumerating objects: 69, done.
remote: Counting objects: 100% (65/65), done.
remote: Compressing objects: 100% (32/32), done.
remote: Total 55 (delta 40), reused 37 (delta 23), pack-reused 0
Unpacking objects: 100% (55/55), 14.77 KiB | 889.00 KiB/s, done.
From https://github.com/janet-lang/spork
   a764ec1..cb4a883  master     -> origin/master
From https://github.com/janet-lang/spork
 * branch            HEAD       -> FETCH_HEAD
HEAD is now at cb4a883 Add *use-rdynamic* and *smart-libs* to control linking.
generating /home/user/.local/lib/janet/.manifests/spork.jdn...
Installed as 'spork'.
copying spork to /home/user/.local/lib/janet...
installing bin/janet-format to /home/user/.local/bin/janet-format
installing bin/janet-netrepl to /home/user/.local/bin/janet-netrepl
copying build/spork/json.so to /home/user/.local/lib/janet/spork...
copying build/spork/json.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/json.a to /home/user/.local/lib/janet/spork...
copying build/spork/rawterm.so to /home/user/.local/lib/janet/spork...
copying build/spork/rawterm.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/rawterm.a to /home/user/.local/lib/janet/spork...
copying build/spork/crc.so to /home/user/.local/lib/janet/spork...
copying build/spork/crc.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/crc.a to /home/user/.local/lib/janet/spork...
copying build/spork/utf8.so to /home/user/.local/lib/janet/spork...
copying build/spork/utf8.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/utf8.a to /home/user/.local/lib/janet/spork...
copying build/spork/tarray.so to /home/user/.local/lib/janet/spork...
copying build/spork/tarray.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/tarray.a to /home/user/.local/lib/janet/spork...
copying src/tarray.h to /home/user/.local/lib/janet/...
copying build/spork/zip.so to /home/user/.local/lib/janet/spork...
copying build/spork/zip.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/zip.a to /home/user/.local/lib/janet/spork...
copying build/spork/cmath.so to /home/user/.local/lib/janet/spork...
copying build/spork/cmath.meta.janet to /home/user/.local/lib/janet/spork...
copying build/spork/cmath.a to /home/user/.local/lib/janet/spork...

$ janet -v
1.30.0-874fd2ab

$ janet
Janet 1.30.0-874fd2ab linux/x64/gcc - '(doc)' for help
repl:1:> (upscope (import spork/cc) :done)
:done
repl:2:> (setdyn :verbose true)
true
repl:3:> (setdyn cc/*visit* cc/visit-execute-if-stale)
<function visit-execute-if-stale>
repl:4:> (cc/search-static-libraries "m" "rt" "dl")
@[]
repl:5:> (cc/compile-and-link-shared "json.so" "src/json.c")
cc -std=c99 -O2 -I/home/user/.local/lib/janet -L/home/user/.local/lib/janet -L/home/user/.local/lib -fPIC -g -DJANET_BUILD_TYPE=develop -c src/json.c -o ./src___json.c.o -pthread
src/json.c:23:10: fatal error: janet.h: No such file or directory
   23 | #include <janet.h>
      |          ^~~~~~~~~
compilation terminated.
error: command failed with non-zero exit code 1
  in os/execute [src/core/os.c] on line 1336
  in visit-execute [/home/user/.local/lib/janet/spork/cc.janet] (tailcall) on line 382, column 7
  in exec [/home/user/.local/lib/janet/spork/cc.janet] (tailcall) on line 74, column 3
  in compile-many [/home/user/.local/lib/janet/spork/cc.janet] on line 210, column 31
  in compile-and-link-shared [/home/user/.local/lib/janet/spork/cc.janet] on line 225, column 26
  in _thunk [repl] (tailcall) on line 5, column 1

$ ls ~/.local/include/janet/janet.h ~/.local/include/janet.h 
/home/user/.local/include/janet.h  /home/user/.local/include/janet/janet.h

$ ls ~/.local/lib/janet/janet.h
ls: cannot access '/home/user/.local/lib/janet/janet.h': No such file or directory
```

**Reply by bakpakin** [reply]
Setting CPATH for the compiler will fix that, however I have added some more code to try and infer include paths from `(dyn *syspath*) .
